### ISSUE TITLE
Performance Improvement Plan

### ISSUE THREAD

**Issue by jpogran** [hashicorp/vscode-terraform#1557]
### Extension Version

v2.27.0

### Problem Statement

We've been investigating performance for terraform-ls to develop a short term and a long term plan to address the issues that have been reported.

We know most of the issues reported surround the work we do in the background to provide features like intellisense, go to references, hover documentation, etc. When you open up the editor, we do a lot of work in the background to understand the code you're working on. We have an indexing process that finds all the Terraform files and modules in your current working directory, parses them, and builds an understanding of all the interdependencies and references inside. We hold this information inside an in-memory database, which we update as you change your files. If you open a directory with many hundreds of folders and files, this may consume more CPU and memory than we intend.

### Proposal

In the short term we're working to add ways to measure and record the performance of terraform-ls in different situations so we can pinpoint the best path forward. We need a way to record the time it takes to perform various tasks in a manner that will give us the information we need while also respecting user privacy and not intruding on usage. We’ve started to instrument opt-in only performance tracing that is currently only used by the developers on the team.

We're also reaching out to users who have reported issues to get a closer look at how they are using VS Code to edit Terraform code. We've noticed that people edit their Terraform code using many different approaches. Some open single files, some open single folders, others open a directory containing many hundreds of modules and Terraform files. No individual approach is wrong or right, but each requires some consideration when developing a language server.

We've implemented several targeted improvements over the last several point releases based on our investigations: ([1332](https://github.com/hashicorp/terraform-ls/pull/1332) and [1258](https://github.com/hashicorp/terraform-ls/pull/1258) ). We’ve also added a language status indicator in [#1547](https://github.com/hashicorp/vscode-terraform/pull/1547) that approximates an indicator when the language server is ready for use so you can see that there are things happening in the background. 

Most of the work we anticipate doing is optimizing these background processes. We plan to use the performance data we collect to inform places where we can either optimize or do less work entirely [#1339](https://github.com/hashicorp/terraform-ls/issues/1339). We are also planning to add a configuration setting  [#1341](https://github.com/hashicorp/terraform-ls/issues/1341) so that users can adjust/tune the amount of workers handling these background jobs to suit their individual needs.

This is by no means all we intend to do, but it is a start. We are hoping these set of steps will inform us on the best approach for resolving these issues long term.

This issue will be pinned to the repo and updated with links to work being done to address this as they are planned and taken on.

### References

- [Open performance issues](https://github.com/issues?q=is%3Aopen+archived%3Afalse+repo%3Ahashicorp%2Fvscode-terraform+repo%3Ahashicorp%2Fterraform-ls+label%3Aperformance)

### Help Wanted

- [ ] I'm interested in contributing a fix myself

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by adrianisk** [comment]
Thanks for the update!

**Comment by jorhett** [comment]
I think some sensible recognition of root modules in large monorepos would greatly help. You don't need to rescan every file (traces show you are) in a huge monorepo while I'm working on a single module.

**Comment by FernandoMiguel** [comment]
I used to have one window with many workspaces and then 1 or 2 windows with a much smaller number of workspaces (usually a couple root modules and the modules they depend on).
Sadly I can't disable terraform on just one window as it seems to be a global plugin.

**Comment by jpogran** [comment]
We've just released `2.27.2` which contains two fixes (https://github.com/hashicorp/terraform-ls/pull/1369, https://github.com/hashicorp/terraform-ls/pull/1372) in our continuing efforts to address this issue.

With the fix, we should be back to `0.29.2` levels of CPU usage while maintaining the memory improvements of `0.29.3`.

If you have the time, please give it a try and let us know how it works for you. Please open a new issue rather than replying here, as we're using this issue as a tracking issue.

**Comment by seanamos** [comment]
I just tested the changes in one of our older, large and monolithic terraform repos. The CPU usage and performance in the root of this workspace was terrible.
I'm seeing a _dramatic_ improvement and both CPU usage and perceived responsiveness.

**Comment by vinayhegde1990** [comment]
I found <ins>#1328</ins> which led me to this central issue for tracking all performance debugging. My current setup has `v2.28.2` of the `hashicorp.terraform` plugin for VSCode `1.83.1` which consumes ~1.4 GB RAM out of 8 GB, below is an image from the OS activity monitor

![terraform-ls-memory-1](https://github.com/hashicorp/vscode-terraform/assets/19818985/d4e5779a-67b2-483c-af51-ba8de219bd09)

A workaround is to enable this plugin only during terraform development and disable it right afterwards. Seems this is not fixed yet, please let me know if any more information is required and if I could help fix this permanently.

**Comment by adamrothman** [comment]
Hi everyone – I'm just checking into say that the performance of format-on-save for Terraform files is absolutely horrendous for me.

I'm working in a very large project with thousands of individual .tf files spread across many environments and modules. Even just changing a single line of a single file, and saving that file, causes the plugin to hang for ~30s, sometimes up to or longer than a minute. This seems insane, given that it takes milliseconds to run `terraform fmt` on a single file. Opening the VSCode console to `Output > HashiCorp Terraform` while this is happening shows thousands of lines like this
```
2023/11/17 09:03:35 jobs.go:461: JOBS: Finishing job "14867": "OpTypePreloadEmbeddedSchema" for {"file:///REDACTED/terraform/modules/vault/modules/basic"} (err = file:///REDACTED/terraform/modules/vault/modules/basic: state not changed, deferredJobs: [])
2023/11/17 09:03:35 jobs.go:321: JOBS: Dispatching next job "14868" (scheduler prio: -1, job prio: 0, isDirOpen: false): "OpTypeDecodeReferenceTargets" for {"file:///REDACTED/terraform/modules/vault/modules/basic"}
2023/11/17 09:03:35 jobs.go:461: JOBS: Finishing job "14868": "OpTypeDecodeReferenceTargets" for {"file:///REDACTED/terraform/modules/vault/modules/basic"} (err = file:///REDACTED/terraform/modules/vault/modules/basic: state not changed, deferredJobs: [])
2023/11/17 09:03:35 jobs.go:321: JOBS: Dispatching next job "14869" (scheduler prio: -1, job prio: 0, isDirOpen: false): "OpTypeDecodeReferenceOrigins" for {"file:///REDACTED/terraform/modules/vault/modules/basic"}
2023/11/17 09:03:35 jobs.go:461: JOBS: Finishing job "14869": "OpTypeDecodeReferenceOrigins" for {"file:///REDACTED/terraform/modules/vault/modules/basic"} (err = file:///REDACTED/terraform/modules/vault/modules/basic: state not changed, deferredJobs: [])
2023/11/17 09:03:35 jobs.go:321: JOBS: Dispatching next job "14870" (scheduler prio: -1, job prio: 0, isDirOpen: false): "OpTypeParseVariables" for {"file:///REDACTED/terraform/modules/vault/modules/basic"}
2023/11/17 09:03:35 jobs.go:461: JOBS: Finishing job "14870": "OpTypeParseVariables" for {"file:///REDACTED/terraform/modules/vault/modules/basic"} (err = file:///REDACTED/terraform/modules/vault/modules/basic: state not changed, deferredJobs: [])
2023/11/17 09:03:35 jobs.go:321: JOBS: Dispatching next job "14871" (scheduler prio: -1, job prio: 0, isDirOpen: false): "OpTypeDecodeVarsReferences" for {"file:///REDACTED/terraform/modules/vault/modules/basic"}
2023/11/17 09:03:35 jobs.go:461: JOBS: Finishing job "14871": "OpTypeDecodeVarsReferences" for {"file:///REDACTED/terraform/modules/vault/modules/basic"} (err = file:///REDACTED/terraform/modules/vault/modules/basic: state not changed, deferredJobs: [])
2023/11/17 09:03:35 jobs.go:321: JOBS: Dispatching next job "14872" (scheduler prio: -1, job prio: 0, isDirOpen: false): "OpTypeParseModuleConfiguration" for {"file:///REDACTED/terraform/modules/vault/modules/basic"}
```

The plugin is doing a ton of unnecessary work every time I save. I understand there's a lot of bookkeeping that needs to happen behind the scenes, but this seems like a scheduling problem to me. IMO the formatting should happen first, with much higher priority than whatever else the plugin needs to do.

I'm running version `v2.29.0` of the `hashicorp.terraform` plugin in VSCode 1.84.2:
```
Version: 1.84.2 (Universal)
Commit: 1a5daa3a0231a0fbba4f14db7ec463cf99d7768e
Date: 2023-11-09T10:52:33.687Z
Electron: 25.9.2
ElectronBuildId: 24603566
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Darwin x64 23.1.0
```

**Comment by jonbackhaus-mt** [comment]
As of v2.29.2, I'm still seeing ~10 GB of RAM usage by `terraform-ls`. Save performance has generally improved (through occasionally hangs).

**Comment by thatguythat1031** [comment]
I'm seeing CPU usage of 200%+ and memory usage of over 9GB (I've seen up to 13Gb) on 2.29.3. I've tried disabling Enhanced Validation and that seems to help, but only a little.

**Comment by jonbackhaus-mt** [comment]
I'll add that enabling `TF_PLUGIN_CACHE_DIR="~/.terraform.d/plugin-cache"` also seems to help performance (though that could be coincidental).

**Comment by thatguythat1031** [comment]
My issue appears to be caused by an internal terraform modules repo we maintain. Many of our other repos use the internal modules repo, which results in every terraform directory in several of our large repos having a copy of that modules repo in its .terraform directory. 

Deleting as many of those .terraform directories as I can brings down the memory usage significantly- I was able to reduce usage from 10gb to 5.6gb by deleting around 8 directories. But that's only a temporary solution and requires constant upkeep.

**Comment by dogmatic69** [comment]
Coming here from https://github.com/hashicorp/terraform-ls/issues/1272, I'd guess my issue is similar to the comment just above. 

> My issue appears to be caused by an internal terraform modules repo we maintain. Many of our other repos use the internal modules repo, which results in every terraform directory in several of our large repos having a copy of that modules repo in its .terraform directory.

running latest stable ubuntu, terraform-ls version seems to be `0.32.8`

Machine has 16Gb ram and terraform-ls process results in system using swap and subsequently becoming completely unresponsive.

**Comment by JBG-brad** [comment]
Running into the same problem with internal modules we maintain.  Working on one of those modules on an M3 Mac Pro, I'm seeing 700% CPU usage and 24GB of RAM.  It's made using the VSCode plugin for terraform completely impossible.

Folder structure is pretty small, just a subfolder for the module and the parent directory.

This came out of nowhere, I've made changes to this module for months and everything has been fine.  Today; it's completely unusable.

The weird thing I'm noting is that my system is making _millions_ of calls to terraform cloud, but we do NOT use it.  Watching the calls via wireshark, there seems to be some sort of infinite recursion.  It appears to be trying to sync to terraform cloud, failing, then trying again at an ever increasing rate.

**Comment by JBG-brad** [comment]
> Hi @JBG-brad, I would love to get to the bottom of this, but so far we haven't been able to reproduce most of the shared cases of extreme resource usage.
> 
> We try to get the schema for public modules from the public registry (https://registry.terraform.io/), but that should only happen once and the result is cached. It shouldn't happen for internal modules, and it definitely shouldn't happen in recursion. So I'm very curious to know what's going on. Maybe this could lead us to a new insight and improve performance for everyone.
> 
> Can you share the full language server log output with me? Or, time permitting, could you join a call and show me the bug in action? You can also [GPG-encrypt](https://keybase.io/dbanck) it if you're worried about sensitive data in it. You can also email it privately to `daniel.banck<at>hashicorp.com` if you prefer.
> 
> If anyone else wants to share their log privately, or has an easily reproducible performance problem, please don't hesitate to contact me.

I'd be happy to jump on a call and show you what's going on.   I believe my email is in my profile.

**Comment by dbanck** [comment]
We've just released `v2.31.2024061114`, which ships with a re-architected language server for improved performance and resource usage ([terraform-ls#1667](https://github.com/hashicorp/terraform-ls/issues/1667)). The language server will now start up much faster and use fewer resources, especially on larger workspaces.

In VS Code, you can opt in to the extension preview:
![CleanShot 2024-06-17 at 12 18 50@2x](https://github.com/hashicorp/vscode-terraform/assets/45985/2e23ba89-e070-4b75-a531-9d3ee3965cdc)

If you have the time, please give it a try and let us know how it works for you. We have tested this with workspaces of different sizes, but still expect some bugs due to the wide variety of configurations. Your feedback is greatly appreciated. Please open a new issue rather than replying here, as we're using this issue as a tracking issue.

**Comment by FernandoMiguel** [comment]
@dbanck i've open all the projects I have in my drive and seems to be behaving much better than ever before
<img width="658" alt="image" src="https://github.com/hashicorp/vscode-terraform/assets/168366/4ed560dc-13f6-41d8-bbc7-b557b45b60d7">
![image](https://github.com/hashicorp/vscode-terraform/assets/168366/7d9b4efc-7a34-4182-9cfa-10c6f76d0a15)

**Comment by dogmatic69** [comment]
I've just installed it and restarted everything. Opened up some of my bigger projects and much better. Comfortably opened 4 projects 🫢 

VS Code is just about snappy again, opening a new project takes a few seconds and does not completely choke up my machine 🚀

**Comment by adamrothman** [comment]
I can confirm that performance is massively improved using the pre-release version – thank you very much to all involved! Awesome work.

**Comment by vinayhegde1990** [comment]
I've also updated to the pre-release version `v2.31.2024061114` and the memory usage is now between 30-40 MB on the OS activity monitor, a thorough 90-95% reduction from the previous version.

![vscode-terraform-new](https://github.com/hashicorp/vscode-terraform/assets/19818985/ba8deef5-9a0f-4ef9-a661-40a9009fc8d9)

thank you very much to everyone involved, great stuff 👏🏻

**Comment by migueleliasweb** [comment]
Sadly, for bigger projects (100k lines), I feel the Terraform extension still tries way too hard and must be looking to parse every single file recursively - this is the only possibility I can think of.

I'm still seeing the language server taking over 1 minute to return even after adding a single character to an existing string.

![image](https://github.com/hashicorp/vscode-terraform/assets/1608212/ebbd1926-8783-44c3-8d27-a5a95bf8eeee)

I can sometimes see CPU spikes of 450%+ and a lot more memory than in this picture :crying_cat_face: .

**Comment by dbanck** [comment]
@migueleliasweb Can you double check if you're using the preview version of the extension? Usually the path to `terraform-ls` should contain the exact version number, in this case `.../extensions/hashicorp.terraform-2.31.2024061114-linux-x64/bin/...`. The path in your screenshot looks like the stable version

**Comment by migueleliasweb** [comment]
Trying now. 

First impressions (30s of it) is that the "time to first interaction" feels faster and it seems to be using less memory.

I'll post if I find anything. :saluting_face: 

I didn't know this thread was specifically abou the preview given the last stable was released 4 days ago. My bad.

![image](https://github.com/hashicorp/vscode-terraform/assets/1608212/69129fec-2904-46e9-bff0-475a744af79d)

**Comment by dogmatic69** [comment]
The original issue was about perfomace, an update in preview was [released 2 weeks ago](https://github.com/hashicorp/vscode-terraform/issues/1557#issuecomment-2172984054)

**Comment by dbanck** [comment]
We've shipped the re-architected language server in the [2.32.0](https://github.com/hashicorp/vscode-terraform/releases/tag/v2.32.0) stable release last week. The language server will (still) start up much faster and use fewer resources, especially on larger workspaces.

Please let us know if you experience any problems or regressions.

**Comment by tyrken** [comment]
Our terraform monorepo has only 1400 files (some of those are symlinked'ed repeats) over 110 folders, but I've have to give up on the VSCode extension two weeks ago.  It took so long to save & format files (20-30 seconds) that I occasionally planned/applied on stale file content instead of the intended updated file, i.e. the extension was causing operational mistakes/bugs.

I fear a lot of "nice to have" features have been added as the expense of performance of the core stuff - syntax highlighting & auto-formatting.  When we can run `terraform fmt -recursive` that checks the whole tree from scratch in less than 4 seconds, how come VSCode is taking so long on warm pre-processed data?

**Comment by seanhoughton** [comment]
I have a fairly small repository and I am frequently unable to save files, sometimes for 10 to 15 minutes, when I have the language server turned on.

There are 85 tf files in the env folder totaling 3,823 lines
There are 76 .tf files in the module folders totaling 4,432 lines

All told there are 977 resources with for_each instances expanded.

During the stalls the log output from the plugin is ripping along with millions of output lines. Here are the occurrences of "ops" in the the last 100k lines in the log:

| Occurrences | Operation                       |
| ----------: | ------------------------------- |
|       14232 | OpTypeDecodeReferenceOrigins    |
|       14231 | OpTypeDecodeReferenceTargets    |
|       14232 | OpTypeDecodeWriteOnlyAttributes |
|         370 | OpTypeGetModuleDataFromRegistry |
|           3 | OpTypeGetTerraformVersion       |
|       12780 | OpTypeLoadModuleMetadata        |
|           3 | OpTypeObtainSchema              |
|       12780 | OpTypeParseModuleConfiguration  |
|           3 | OpTypeParseModuleManifest       |
|           3 | OpTypeParseProviderVersions     |
|           3 | OpTypeParseTerraformSources     |
|       14230 | OpTypePreloadEmbeddedSchema     |
|         324 | OpTypeReferenceValidation       |
|         324 | OpTypeSchemaModuleValidation    |


Even just turning down the logging verbosity might help at this point.

**Comment by karoldrazek** [comment]
Dear Team,

is this being worked on at all? I am honestly asking.

**Comment by Danixu** [comment]
I'm facing the same issue (that is why I have endede here).

On my case It takes almost forever. For example, I had time to search on Google for the issue, visit two or three posts until some matched, read some comments on that post until I have found the one that points me to this issue, read the entire issue and even at this point the Intelisense formatter still working without to save the file.

A summary can be that I have to close the entire Code program and reopen it to be able to save the files (luckily the changes are kept), because only when it is recently opened works to save a file. Another workaround is to disable the auto formatting on save.

Best regards.

**Comment by Poltergeisen** [comment]
seeing tons of performance issues on medium sized repos. Not sure if virus scanners are impacting this or not, but it'd be nice to be able to track this somehow.

**Comment by 12345ieee** [comment]
Performance has been terrible for so many releases, and communication on this issue seems completely absent.

What I can recommend to anyone reading is to downgrade to v2.27.2 and keep it there.
That version lacks support for some minor recent syntax changes, but it's fast and usable.

**Comment by rgegriff** [comment]
FWIW; I had similar issues; but I just disabled "Format on save" and "enhanced validation"; and it is much MUCH better. I can actually type without my whole machine hanging

**Comment by Conzar** [comment]
I am experiencing a large spike in memory usage after opening several directories that contain terraform modules.

Based on the problem statement that terraform-ls uses an in memory database, could a file based database like [SQLite](https://sqlite.org/) be a solution?
